### ISSUE TITLE
dynamic requirements: what's wrong with what I am doing?

### ISSUE THREAD

**Issue by tgy** [spotify/luigi#3255]
here's a dead-simple self-contained example of what i'm trying to do in a much more complex workflow

if you run the following code, you'll see that it never terminates.

i think there's something i'm doing wrong with the luigi API but i'm not sure what's the issue

```python
import luigi.local_target

import luigi


class DatumTask(luigi.Task):
    split = luigi.Parameter(significant=False)

    def output(self):
        return luigi.local_target.LocalTarget(f"/tmp/{self.split}.bin")

    def run(self):
        with self.output().open("w"):
            pass


class DatasetTask(luigi.Task):
    def output(self):
        return luigi.local_target.LocalTarget("/tmp/_SUCCESS")

    def run(self):
        for split in ["train", "test", "valid"]:
            yield DatumTask(split=split)
        with self.output().open("w"):
            pass


task = DatasetTask()
luigi.build([task], local_scheduler=True)
```

my observation is that if i execute the script once, it creates `train.bin` and `test.bin` successfully, and then it keeps running forever, never terminating

if i interrupt the run, and then run it again, it creates `valid.bin`, validates `DatasetTask` as complete, and correctly terminates

my issue is that in my real-world pipeline i have many more tasks, and i don't want to restart the script several times for the workflow to terminate.

how can i fix this?

**Comment by tgy** [comment]
I tried using `DynamicRequirements` but it's not solving my issue

tagging @riga because he might know what the issue is

**Comment by riga** [comment]
Hi @tgy ,

your split parameter is not significant, resulting in the luigi scheduler not triggering all three tasks but just a subset of tasks (in your case one or two of them; I think this should be only one though). As a result your second task declares dynamic dependencies that are never run and thus it never makes it to the point where its own run method finishes.

The DynamicDependencies container is just a way of saving resources during repeated completness checks, e.g. in cases with large numbers of dynamic deps with remote targets.

So, in short, setting significant back to True should do the trick.

Hope that helps!

**Comment by tgy** [comment]
thanks @riga, that was exactly it. i figured it out by myself before your message but did not really understand why making the parameter significant solved it. now i got it :) thanks a lot!
